src/imitation/data/wrappers.py

```diff
@@ -87,16 +98,42 @@ def _finish_partial_trajectories(self) -> Sequence[types.TrajectoryWithRew]:
                 self._traj_accum.add_step({"obs": traj.obs[-1]}, key=i)
         return trajs
 
-    def pop_trajectories(self) -> Sequence[types.TrajectoryWithRew]:
-        """Pops recorded trajectories as a list of TrajectoryWithRew instances."""
-        if self.n_transitions == 0:
-            return []
-        partial_trajs = self._finish_partial_trajectories()
-        self._trajectories.extend(partial_trajs)
+    def pop_finished_trajectories(
+        self,
+    ) -> Tuple[Sequence[types.TrajectoryWithRew], Sequence[int]]:
+        """Pops recorded complete trajectories `trajs` and episode lengths `ep_lens`.
+
+        Returns:
+            A tuple `(trajs, ep_lens)` where `trajs` is a sequence of trajectories
+            including the terminal state (but possibly missing initial states, if
+            `pop_trajectories` was previously called) and `ep_lens` is a sequence
+            of episode lengths. Note the episode length will be longer than the
+            trajectory length when the trajectory misses initial states.
+        """
         trajectories = self._trajectories
+        ep_lens = self._ep_lens
         self._trajectories = []
+        self._ep_lens = []
         self.n_transitions = 0
-        return trajectories
+        return trajectories, ep_lens
+
+    def pop_trajectories(
+        self,
+    ) -> Tuple[Sequence[types.TrajectoryWithRew], Sequence[int]]:
+        """Pops recorded trajectories `trajs` and episode lengths `ep_lens`.
+
+        Returns:
+            A tuple `(trajs, ep_lens)`. `trajs` is a sequence of trajectory fragments,
+            consisting of data collected after the last call to `pop_trajectories`.
+            They may miss initial states (if `pop_trajectories` previously returned
+            a fragment for that episode), and terminal states (if the episode has
+            yet to complete). `ep_lens` is the total length of completed episodes.
+        """
+        if self.n_transitions == 0:
+            return [], []
+        partial_trajs = self._finish_partial_trajectories()
+        self._trajectories.extend(partial_trajs)
+        return self.pop_finished_trajectories()
```

Comment:
If I understand it correctly, it's possible for both `pop_trajectories()` and `pop_finished_trajectories()` to have incomplete trajectories. The difference being that 
- All trajs from `pop_finished_trajectories()` are terminated. There can be one without initial state.
- `pop_trajectories()`: There can be one without initial state or non-terminated yet.

Reply:
Yes, that sounds right. But I think the idea is that if you _always_ use `pop_finished_trajectories()`, then you'll only get complete trajectories.